src/probability_theory/density.lean

```diff
@@ -0,0 +1,406 @@
+  {g : E → F} (hg : quasi_measure_preserving g μ ν)
+  (hmap : (map g (map X ℙ)).have_lebesgue_decomposition ν) :
+  has_pdf (g ∘ X) ℙ ν :=
+begin
+  rw [has_pdf_iff (hg.measurable.comp hX), ← map_map hg.measurable hX],
+  refine ⟨hmap, _⟩,
+  rw [map_eq_with_density_pdf X ℙ μ],
+  refine absolutely_continuous.mk (λ s hsm hs, _),
+  rw [map_apply hg.measurable hsm, with_density_apply _ (hg.measurable hsm)],
+  have := hg.absolutely_continuous hs,
+  rw map_apply hg.measurable hsm at this,
+  exact set_lintegral_measure_zero _ _ this,
+end
+
+lemma quasi_measure_preserving_has_pdf' [is_finite_measure ℙ] [sigma_finite ν]
+  {X : α → E} (hX : measurable X) [has_pdf X ℙ μ]
+  {g : E → F} (hg : quasi_measure_preserving g μ ν) :
+  has_pdf (g ∘ X) ℙ ν :=
+begin
+  haveI : is_finite_measure (map g (map X ℙ)) :=
+    @is_finite_measure_map _ _ _ _ (map X ℙ) (is_finite_measure_map ℙ hX) _ hg.measurable,
+  exact quasi_measure_preserving_has_pdf hX hg infer_instance,
+end
+
+end
+
+section real
+
+variables [is_finite_measure ℙ] {X : α → ℝ} (hX : measurable X)
+
+include hX
+
+/-- A real-valued random variable `X` `has_pdf X ℙ λ` (where `λ` is the Lebesgue measure) if and
+only if the push-forward measure of `ℙ` along `X` is absolutely continuous with respect to `λ`. -/
+lemma real.has_pdf_iff : has_pdf X ℙ ↔ map X ℙ ≪ volume :=
+begin
+  haveI : is_finite_measure ((map X) ℙ) := is_finite_measure_map ℙ hX,
+  rw [has_pdf_iff hX, and_iff_right_iff_imp],
+  exact λ h, infer_instance,
+end
+
+/-- If `X` is a real-valued random variable that has pdf `f`, then the expectation of `X` equals
+`∫ x, x * f x ∂λ` where `λ` is the Lebesgue measure. -/
+lemma integral_mul_eq_integral [has_pdf X ℙ]:
+  ∫ x, x * (pdf X ℙ volume x).to_real ∂(volume) = ∫ x, X x ∂ℙ :=
+integral_mul_eq_integral' hX measurable_id
+
+lemma has_finite_integral_mul {f : ℝ → ℝ} {g : ℝ → ℝ≥0∞}
+  (hg : pdf X ℙ =ᵐ[volume] g) (hgi : ∫⁻ x, ∥f x∥₊ * g x ∂(volume) < ∞) :
+  has_finite_integral (λ x, f x * (pdf X ℙ volume x).to_real) volume :=
+begin
+  rw [has_finite_integral],
+  have : (λ x, ↑∥f x∥₊ * g x) =ᵐ[volume] (λ x, ∥f x * (pdf X ℙ volume x).to_real∥₊),
+  { refine ae_eq_trans (filter.eventually_eq.mul (ae_eq_refl (λ x, ∥f x∥₊))
+      (ae_eq_trans hg.symm (of_real_to_real_ae_eq hX).symm)) _,
+    simp_rw [← smul_eq_mul, nnnorm_smul, ennreal.coe_mul, smul_eq_mul],
+    refine filter.eventually_eq.mul (ae_eq_refl _) _,
+    convert ae_eq_refl _,
+    ext1 x,
+    exact real.ennnorm_eq_of_real ennreal.to_real_nonneg },
+  rwa ← lintegral_congr_ae this,
+end
+
+end real
+
+section
+
+/-! **Uniform Distribution** -/
+
+/-- A random variable `X` has uniform distribution if it has a probability density function `f`
+with compact support `s` such that `f = (μ s)⁻¹ 1ₛ` a.e. where `1ₛ` is the indicator function
+for `s`.
+
+We require compact support since otherwise the expectation might not be finite. -/
+class uniform {m : measurable_space α} (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac)
```

Comment:
Suggested by Scott: make support a parameter
Suggested by Remy: make it into a def

Reply:
I'm in favour of making it a `def`. It can be switched to a `class` later if there are concrete examples where this helps.

Reply:
Is this definition what you had in mind? And then we can add additional conditions on the support afterward.
```lean
def is_uniform {m : measurable_space α} (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac)
  (support : set E) :=
pdf X ℙ μ =ᵐ[μ] support.indicator ((μ support)⁻¹ • 1)
```

Reply:
Yes.